src/main/scala/GenericFIR.scala

```diff
@@ -0,0 +1,177 @@
+
+class GenericFIRCellBundle[T<:Data:Ring](genIn:T, genOut:T) extends Bundle {
+  val data: T = genIn.cloneType
+  val carry: T = genOut.cloneType
+
+  override def cloneType: this.type = GenericFIRCellBundle(genIn, genOut).asInstanceOf[this.type]
+}
+object GenericFIRCellBundle {
+    def apply[T<:Data:Ring](genIn:T, genOut:T): GenericFIRCellBundle[T] = new GenericFIRCellBundle(genIn, genOut)
+}
+
+class GenericFIRCellIO[T<:Data:Ring](genIn:T, genOut:T) extends Bundle {
+    val coeff = Input(genIn.cloneType)
+    val in = Flipped(Decoupled(GenericFIRCellBundle(genIn, genOut)))
+    val out = Decoupled(GenericFIRCellBundle(genIn, genOut))
+}
+object GenericFIRCellIO {
+    def apply[T<:Data:Ring](genIn:T, genOut:T): GenericFIRCellIO[T] = new GenericFIRCellIO(genIn, genOut)
+}
+
+class GenericFIRBundle[T<:Data:Ring](proto: T) extends Bundle {
+  val data: T = proto.cloneType
+
+  override def cloneType: this.type = GenericFIRBundle(proto).asInstanceOf[this.type]
+}
+object GenericFIRBundle {
+  def apply[T<:Data:Ring](proto: T): GenericFIRBundle[T] = new GenericFIRBundle(proto)
+}
+
+class GenericFIRIO[T<:Data:Ring](genIn:T, genOut:T) extends Bundle {
+    val in = Flipped(Decoupled(GenericFIRBundle(genIn)))
+    val out = Decoupled(GenericFIRBundle(genOut))
+}
+object GenericFIRIO {
+    def apply[T<:Data:Ring](genIn:T, genOut:T): GenericFIRIO[T] = new GenericFIRIO(genIn, genOut)
+}
+
+// A generic FIR filter
+class GenericFIR[T<:Data:Ring](genIn:T, genOut:T, coeffs: Seq[T]) extends Module {
+  val io = IO(GenericFIRIO(genIn, genOut))
+
+  // Construct a vector of GenericFIRDirectCells
+  val directCells = Seq.fill(coeffs.length){ Module(new GenericFIRDirectCell(genIn, genOut)).io }
+
+  // connect coefficients
+  for ((cell, coeff) <- directCells.zip(coeffs)) {
+    cell.coeff := coeff
+  }
+
+  // connect input to first cell.
+  directCells(0).in.bits.data := io.in.bits
+  directCells(0).in.bits.carry := Ring[T].zero // first cell has no carry
+  directCells(0).in.valid := io.in.valid
+  io.in.ready := directCells(0).in.ready
+
+  // connect adjacent cells
+  for ((prev, next) <- directCells.zip(directCells.tail)) {
+    next.in.bits := prev.out.bits
+    next.in.valid := prev.out.valid
+    prev.out.ready := next.out.ready
+  }
+
+  // connect output to last cell
+  io.out.bits := directCells.last.out.bits.carry
+  io.out.valid := directCells.last.out.valid
+  directCells.last.out.ready := io.out.ready
+}
+
+// A generic FIR direct cell used to construct a larger direct FIR chain
+//
+//   in ----- [z^-1]-- out
+//              |
+//   coeff ----[*]
+//              |
+//   carryIn --[+]-- carryOut
+//
+class GenericFIRDirectCell[T<:Data:Ring](genIn: T, genOut: T) extends Module {
+  val io = IO(GenericFIRCellIO(genIn, genOut))
+
+  // Passthrough ready
+  io.in.ready := io.out.ready
+
+  // TODO reason about ready/valid
+  // Describe what's going on
+  val validReg = RegNext(io.in.fire(), init = false.B)
+  io.out.valid := validReg
```

Comment:
I'm not sure that my logical changes are right, you should make sure ready/valid work